File: kernel/trap.c
// SPDX-License-Identifier: GPL-3.0-or-later
/*
 * SlipperKernel — trap dispatch.
 *
 * scause (top bit = interrupt):
 *   0x8000000000000005  S-mode timer interrupt  -> timer_handle()
 *   0x8000000000000009  S-mode external         -> plic_handle() (TODO)
 *   8                   ecall from U-mode       -> syscall_handler()
 *   2,3,5,7,12,13,15    fault                   -> vmm fault / panic
 */
#include "types.h"
#include "trap.h"
#include "klog.h"
#include "riscv.h"
#include "syscall.h"
#include "timer.h"
#include "vmm.h"
#include "proc.h"
#include "plic.h"

void trap_init(void)
{
    /* stvec = direct mode: | mode(2) | base(>>2) | */
    usize vec = (usize)&trap_entry;
    csr_write(stvec, vec);
    kinf("trap: stvec=0x%lx", vec);
}

void trap_handler(trap_frame_t *f)
{
    u64 cause = csr_read(scause);
    u64 is_int = cause & SCAUSE_INT;
    u64 code   = cause & SCAUSE_CODE_MASK;

    if (is_int) {
        switch (code) {
        case 5: /* S-timer */
            timer_handle();
            break;
        case 9: /* S-external (PLIC) */
            {
                int irq = plic_claim();
                if (irq) {
                    kinf("trap: PLIC IRQ %d", irq);
                    plic_complete(irq);
                }
            }
            break;
        case 1: /* S-soft */
            kwrn("trap: soft IRQ (unhandled)");
            break;
        default:
            kwrn("trap: unknown IRQ scause=0x%lx", cause);
            break;
        }
    } else {
        switch (code) {
        case CAUSE_U_ECALL:
            syscall_handler(f);
            f->sepc += 4;
            break;

        case CAUSE_INST_PF:
        case CAUSE_LD_PF:
        case CAUSE_ST_PF:
        case CAUSE_IAMISS:
        case CAUSE_LDAMISS:
        case CAUSE_STAMISS:
            kerr("trap: page fault sepc=0x%lx stval=0x%lx code=%lu",
                 f->sepc, csr_read(stval), code);
            proc_exit(proc_current()->pid, (int)(code + 100));
            break;

        case CAUSE_ILL:
            kerr("trap: illegal instruction sepc=0x%lx stval=0x%lx",
                 f->sepc, csr_read(stval));
            proc_exit(proc_current()->pid, 132);
            break;

        case CAUSE_BRK:
            kerr("trap: breakpoint sepc=0x%lx", f->sepc);
            proc_exit(proc_current()->pid, 133);
            break;

        default:
            kpanic("trap: unhandled scause=0x%lx sepc=0x%lx stval=0x%lx",
                   cause, f->sepc, csr_read(stval));
        }
    }

    /* Post-dispatch: reschedule if needed. */
    if (proc_current()) {
        if (unlikely(proc_current()->state == PROC_STATE_EXITED)) {
            sched_yield(f);
            khalt();
        }
        if (need_resched) {
            need_resched = 0;
            sched_yield(f);
        }
    }
}

File: kernel/src/srv/trap.rs
//! Trap dispatch.
use crate::arch::regs::*;
use crate::arch::trap_frame::TrapFrame;
use crate::drivers::plic;
use crate::proc;
use crate::srv::timer;
use crate::syscall::handler;

pub unsafe fn init() {
    crate::arch::csr::write_stvec(crate::arch::asm::trap_entry as *const () as usize as u64);
    let stack_top = &crate::arch::__stack_top as *const u8 as usize;
    crate::arch::csr::write_sscratch(stack_top as u64);
    crate::kinf!(
        "trap",
        "stvec=%p",
        onyx_core::fmt::Arg::from(crate::arch::asm::trap_entry as *const () as usize as u64)
    );
}

pub unsafe fn handle(tf: &mut TrapFrame) {
    let scause = crate::arch::csr::read_scause();
    let is_int = scause & SCAUSE_INT != 0;
    let code = scause & !SCAUSE_INT;
    if is_int {
        match code {
            INTR_S_TIMER => timer::handle(),
            INTR_S_EXTERN => {
                plic::dispatch();
            }
            INTR_S_SOFT => {
                crate::kwrn!("trap", "unhandled S-soft interrupt");
            }
            _ => {
                crate::kwrn!(
                    "trap",
                    "unhandled interrupt: code=%d",
                    onyx_core::fmt::Arg::from(code)
                );
            }
        }
    } else {
        match code {
            CAUSE_U_ECALL => {
                let ret = handler::handle(tf);
                tf.a0 = ret as u64;
                tf.sepc = tf.sepc.wrapping_add(4);
            }
            CAUSE_INST_PF | CAUSE_LD_PF | CAUSE_ST_PF | CAUSE_IAMISS | CAUSE_LDAMISS
            | CAUSE_STAMISS => {
                let pid = proc::current_pid();
                let stval = crate::arch::csr::read_stval();
                let sstatus = crate::arch::csr::read_sstatus();
                let from_kernel = sstatus & SSTATUS_SPP != 0;
                if from_kernel || pid == 0 {
                    crate::kerr!(
                        "trap",
                        "KERNEL page fault sepc=%p stval=%p sstatus=%p",
                        onyx_core::fmt::Arg::from(tf.sepc),
                        onyx_core::fmt::Arg::from(stval),
                        onyx_core::fmt::Arg::from(sstatus)
                    );
                    crate::srv::klog::halt();
                }
                crate::kerr!(
                    "trap",
                    "page fault pid=%d sepc=%p stval=%p",
                    onyx_core::fmt::Arg::from(pid),
                    onyx_core::fmt::Arg::from(tf.sepc),
                    onyx_core::fmt::Arg::from(stval)
                );
                proc::exit(pid, 100 + code as i32);
            }
            CAUSE_ILL => {
                let pid = proc::current_pid();
                crate::kerr!(
                    "trap",
                    "illegal instruction pid=%d sepc=%p",
                    onyx_core::fmt::Arg::from(pid),
                    onyx_core::fmt::Arg::from(tf.sepc)
                );
                proc::exit(pid, 132);
            }
            CAUSE_BRK => {
                let pid = proc::current_pid();
                proc::exit(pid, 133);
            }
            _ => {
                crate::kpanic!(
                    "trap",
                    "unhandled exception: scause=%p sepc=%p",
                    onyx_core::fmt::Arg::from(scause),
                    onyx_core::fmt::Arg::from(tf.sepc)
                );
            }
        }
    }
    // Signal delivery: check the current process for pending unblocked
    // signals. KILL terminates the process; other signals are cleared (MVP).
    proc::signal_check(tf);
    let pid = proc::current_pid();
    if pid != 0 {
        if let Some(p) = proc::by_pid(pid) {
            if matches!(p.state, proc::ProcState::Exited) {
                proc::sched_yield(tf);
                crate::srv::klog::halt();
            }
        }
    }
    if proc::NEED_RESCHED {
        proc::sched_yield(tf);
    }
}
